### ISSUE TITLE
Prefetch isn't executed when running locally

### ISSUE THREAD

**Issue by daften** [qltysh/codeclimate#805]
Latest codeclimate installed. No output about prefetch in the debug output

I expect my files to be prefetched, but they aren't. The repo in question is a private repository, but my codeclimate file is here: https://gist.github.com/daften/9a884ff729b69d0a63e4e2c85d5f771a

I run on macosX 10.13.2, docker is installed through brew, as is codeclimate. If needed, I can provide a copy of the code.

**Comment by ABaldwinHunter** [comment]
Hi @daften! Thanks for writing in.

I believe to get those files locally, you'll need to run the `codeclimate prepare` step separately:

```
$ codeclimate prepare
$ codeclimate analyze
```

It's a bit hidden, but in `codeclimate help` output there's a description:
```
    prepare [--allow-internal-ips]                      Run the commands in your prepare step.
```

Can you try that out and let me know how it goes?

Thanks!

Ashley

**Comment by daften** [comment]
Hi,

I completely looked over that, my apologies. That does work, thank you for that!

However when I discussed this with some of our developers, we all felt it would be more natural if the analyze step would also fetch (and default delete afterwards) the config files in the prefetch step. A flag could indicate not to do this, e.g. when you want to try to change those files and are testing locally in the repository.

I don't know if this could be accepted as a feature request?

thanks!

**Comment by wfleming** [comment]
Hi @daften,

Thanks for the feedback. One reason we don't want to `fetch` to be done automatically every time analysis runs is that one useful use-case for the CLI is working offline, so relying on a network connection seems like an anti-pattern, and would make analysis slower to run as well. We also generally try to design the CLI as sets of small, narrowly focused, indepenent commands that can be composed as necessary and don't cross-run each other too much.

That being said, I realize the workflow can be a bit confusing as-is since nothing clearly says `prepare` needs to be run first. I think our docs could improve in this area, but as a functionality change I also think it would be reasonable to have `analyze` also run `prepare` when it detects the config file `prepare` will write are missing locally, which I think would address the biggest pain point.

> and default delete afterwards

Can you comment a bit on why that's a desirable workflow for you? In most team's usage of this feature that we've seen, they change fairly rarely, so always fetching and deleting seems wasteful. If they're annoying because they show up in diffs or something, we do recommend adding files fetched via `prepare` to your `.gitignore`.

**Comment by daften** [comment]
Hi Will,

The use case of working purely offline is something we hadn't thought
about, and it would slow it down a bit, although that last part shouldn't
be a big issue nowadays. Then again, we're privileged in our work
circumstances.

The fix you suggest, to download if the local files are missing, seem like
a very good way to work with it. Documentation improvements are also
possible. Something simple like outputting on analyze that the config files
are only fetched with prepare would also be a good solution for us.

The default delete afterwards is because I was thinking about getting local
setup as close to codeclimate cloud as possible. In the cloud, if I'm not
mistaken, there's always the start from a "clean" start. So a fresh clone
or a git clean -fd or something. So the files are fetched every time.
Deleting the files afterwards seemed like a good solution at the time, but
given your input regarding offline work mostly, scratch that thought.

In conclusion, there's two things that you suggest that seem very feasible
to me and would help us (and most devs I think) tremendously:

   - Fetching the files if they are not present.
   - Improving documentation, with the added item I suggest to output a
   warning or info message on analyze about the prepare step.

Thanks for getting back to us so quickly each time and evaluating this
feature request :)

Dieter



<https://mailtrack.io/> Sent with Mailtrack
<https://chrome.google.com/webstore/detail/mailtrack-for-gmail-inbox/ndnaehgpjlnokgebbaldlmgkapkpjkkb?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality>



--
Dieter Blomme

**Comment by mteodori** [comment]
it would be nice if there was a command line option for analyze to fetch

**Comment by brynary** [comment]
Hello -- Thank you for this GitHub issue.

We are going to be releasing a new version of our code quality CLI shortly, and so I am in the process of closing all open GitHub issues in preparation for that. We will be in touch soon and would be happy to hear from you on any feedback you have on the new CLI when it is available later this month.
